File: src/main/java/com/severett/kotlinintro/model/Pet.java
package com.severett.kotlinintro.model;

import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.MappedSuperclass;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import static jakarta.persistence.GenerationType.IDENTITY;

@MappedSuperclass
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
public abstract class Pet {
    @Id
    @GeneratedValue(strategy = IDENTITY)
    private int id;

    private String name;

    @SuppressWarnings("unused")
    public abstract String getSound();
}

File: src/main/kotlin/com/severett/kotlinintro/model/Pet.kt
package com.severett.kotlinintro.model

import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.MappedSuperclass

@MappedSuperclass
sealed class Pet(
    @field:Id
    @field:GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Int,
    var name: String
) {
    abstract val sound: String
}
